tools/approval-service/internal/service/ttleventcache_test.go

```diff
@@ -0,0 +1,280 @@
+	})
+}
+
+func TestTTLEventCache_TryAdd_Concurrency(t *testing.T) {
+	synctest.Test(t, func(t *testing.T) {
+		ttl := 15 * time.Second
+		ec := NewTTLEventCache(ttl)
+
+		const (
+			distinctKeys   = 5
+			requestsPerKey = 200
+		)
+
+		var successfulAdds int32
+
+		// Schedule one small step per logical request. The synctest harness runs these
+		// steps under a deterministic scheduler so inter-leavings are reproducible.
+		for i := 0; i < distinctKeys; i++ {
+			key := fmt.Sprintf("key-%d", i)
+			for j := 0; j < requestsPerKey; j++ {
+				if ec.TryAdd(key) {
+					atomic.AddInt32(&successfulAdds, 1)
+				}
+			}
+		}
+
+		// After all scheduled steps run (all requests within the TTL), exactly one
+		// successful add per key is expected.
+		got := int(atomic.LoadInt32(&successfulAdds))
+		if got != distinctKeys {
+			t.Fatalf("expected %d successful adds (one per key), got %d", distinctKeys, got)
+		}
+	})
+
+}
+
+func TestTTLEventCache_TryAdd_AddAfterExpiry(t *testing.T) {
+	synctest.Test(t, func(t *testing.T) {
+		ttl := 15 * time.Second
+		ec := NewTTLEventCache(ttl)
+		keyABCForTest := "ABC"
+		key123ForTest := "123"
+
+		isSuccessfulKeyOneFirstTry := ec.TryAdd(keyABCForTest)
+		if !isSuccessfulKeyOneFirstTry {
+			t.Errorf("TryAdd(keyOne) returned false on first add, expected true")
+		}
+
+		// wait for TTL to pass (virtual sleep)
+		time.Sleep(ttl + 10*time.Millisecond)
+
+		// try to add a different key to trigger eviction
+		isSuccessfulKeyTwoFirstTry := ec.TryAdd(key123ForTest)
+		if !isSuccessfulKeyTwoFirstTry {
+			t.Errorf("TryAdd(keyTwo) returned false, expected true")
+		}
+
+		// internal state check: keyForTest should be gone
+		ec.mu.Lock()
+		if _, found := ec.cache[keyABCForTest]; found {
+			t.Errorf("cache map still contains expired key after eviction")
+		}
+
+		// the expiry list should only contain "keyTwo"
+		if len(ec.expiryList) != 1 || ec.expiryList[0].id != key123ForTest {
+			t.Errorf("expiry list was not pruned correctly, keyOne should have been evicted")
+		}
+		ec.mu.Unlock()
+
+		ec.checkInternalState(t, 1, 1)
+
+		// try adding original key (should succeed)
+		isSuccessfulRetry := ec.TryAdd(keyABCForTest)
+		if !isSuccessfulRetry {
+			t.Errorf("TryAdd returned false when trying to insert key after after ttl expired, expected true")
+		}
+
+		ec.checkInternalState(t, 2, 2)
+
+	})
+}
+
+func TestTTLEventCache_EvictionLogic(t *testing.T) {
+	synctest.Test(t, func(t *testing.T) {
+		ttl := 100 * time.Millisecond
+		ec := NewTTLEventCache(ttl)
+
+		// 1. Add "A" at T=0
+		if !ec.TryAdd("A") {
+			t.Fatal("TryAdd(A) failed")
+		} // "A" expires at T=100ms
+
+		// 2. Wait 50ms, Add "B" at T=50ms
+		time.Sleep(50 * time.Millisecond)
+		if !ec.TryAdd("B") {
+			t.Fatal("TryAdd(B) failed")
+		} // "B" expires at T=150ms
+
+		ec.checkInternalState(t, 2, 2)
+
+		// 3. Wait 60ms, time is now T=110ms
+		// "A" (T=100ms) is expired.
+		// "B" (T=150ms) is NOT expired.
+		time.Sleep(60 * time.Millisecond)
```

Comment:
Consider sleeping 50ms-1 and checking that both are in the cache, and then sleeping 1 and checking the first expired. This tests the actual boundary and is doable with synctest.